lib/spack/spack/solver/concretize.lp

```diff
@@ -674,11 +681,33 @@ activate_requirement(node(ID, Package), X) :-
   requirement_conditional(Package, X, Y).
 
 requirement_group_satisfied(node(ID, Package), X) :-
-  1 { condition_holds(Y, node(ID, Package)) : requirement_group_member(Y, Package, X) } 1,
+  1 { requirement_condition_satisfied(Y, node(ID, Package)) : requirement_group_member(Y, Package, X) } 1,
   requirement_policy(Package, X, "one_of"),
   activate_requirement(node(ID, Package), X),
   requirement_group(Package, X).
 
+requirement_condition_satisfied(Y, node(VirtualID, Virtual)) :-
+  condition_holds(Y, node(ProviderID, Provider)),
+  requirement_group_member(Y, Virtual, X),
+  provider(node(ProviderID, Provider), node(VirtualID, Virtual)).
+
+requirement_condition_satisfied(Y, node(ID, Package)) :-
+  condition_holds(Y, node(ID, Package)),
+  requirement_group_member(Y, Package, X).
+
+{ impose(EffectID, node(ID, Package)) } :-
+  pkg_fact(Package, condition_effect(Y, EffectID)),
+  requirement_group_member(Y, Package, X),
+  activate_requirement(node(ID, Package), X),
+  requirement_group(Package, X).
+
+{ impose(EffectID, node(ProviderID, Provider)) } :-
+  pkg_fact(Provider, condition_effect(Y, EffectID)),
+  provider(node(ProviderID, Provider), node(VirtualID, Virtual)),
+  requirement_group_member(Y, Virtual, X),
+  activate_requirement(node(VirtualID, Virtual), X),
+  requirement_group(Virtual, X).
```

Comment:
I'll look at this better, but I don't immediately see why we can give a choice to impose something without having anything that says that the corresponding trigger condition is satisfied in the body of the rule. 

The only explanation I see is that rules in requirements are always of the kind:
```
%----------------------------------------------------------------------------
% Trigger conditions
%----------------------------------------------------------------------------
pkg_fact("openmpi",trigger_id(0)).
pkg_fact("openmpi",trigger_msg("openmpi%gcc+static")).
condition_requirement(0,"node","openmpi").
condition_requirement(0,"variant_value","openmpi","static","True").
condition_requirement(0,"node_compiler","openmpi","gcc").
condition_requirement(0,"node_compiler_version_satisfies","openmpi","gcc",":").

%----------------------------------------------------------------------------
% Imposed requirements
%----------------------------------------------------------------------------
pkg_fact("openmpi",effect_id(0)).
pkg_fact("openmpi",effect_msg("openmpi%gcc+static")).
imposed_constraint(0,"node","openmpi").
imposed_constraint(0,"variant_set","openmpi","static","True").
imposed_constraint(0,"node_compiler_set","openmpi","gcc").
imposed_constraint(0,"node_compiler_version_satisfies","openmpi","gcc",":").
```
but that also means they work logically backwards with this choice - we give choice to impose the rules because they are designed in such a way that imposing the effect implies the causes.

Reply:
> I don't immediately see why we can give a choice to impose something without having anything that says that the corresponding trigger condition is satisfied in the body of the rule.

In this case there is no "valid" corresponding trigger condition:

* the condition requirements generated (the ones you list in the example directly above) are not actually for triggering the impose, but rather to record whether requirement group members are satisfied (technically they also trigger the impose, but they are effectively useless for that purpose)
* currently the only "influence" that the requirement group creates are error conditions when their members are not satisfied
* "somehow", we have a means to thwart error facts by deriving attributes for all requirements other than cflags
  * IMO, it is not worth understanding how or why this works, and saying that we are allowed to impose things is if anyone would want it is reasonable